### PR TITLE
soffice: Implement methods for status bar announcement

### LINKED ISSUE #11698: LibreOffice Writer: Statusline is not announced
### Steps to reproduce:
1. Open Writer
2. Request announcement of status line with screenreader

### Actual behavior:
The screenreader announces  "No informatiion in status line".

### Expected behavior:
status line is announced

#### NVDA installed/portable/running from source:
installed
#### NVDA version:
2020.2

#### Windows version:
Edition	Windows 	10 Home
Version				1909
Operating system build	18363.1016

#### Name and version of other software in use when reproducing the issue:
LibreOffice
Version:	7.0.0.3 (x64)
Build ID:	8061b3e9204bef6b321a21033174034a5e2ea88e
CPU-Threads: 	4; BS: Windows 10.0 Build 18363; UI-Render: Skia/Vulkan; VCL: win
Locale: 	de-DE (de_DE); UI: de-DE
Calc: 	threaded

#### Other information about your system:
---
#### Does the issue still occur after restarting your computer?
Yes.

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
No.

#### If addons are disabled, is your problem still occuring?
Not tested.

#### Did you try to run the COM registry fixing tool in NVDA menu / tools?
Not tested.

### PR DESCRIPTION
**Pull request by michaelweghorn**
### Link to issue number:

Fixes #11698

### Summary of the issue:

Announcement of the status bar in LibreOffice (e.g. triggered by NVDA+End) didn't work because the default implementations do not work for the LibreOffice case.

### Description of user facing changes

Announcement of the status bar (e.g. triggered
by NVDA+End) works for LibreOffice.

### Description of development approach

Override the default app module implementation to retrieve the status bar and announce its text in the LibreOffice app module, since the default implementations are not sufficient for the LibreOffice case:

* To retrieve the status bar object, locate the descendant with the corresponding role, but only descend into children of role `ROOTPANE` and `WINDOW` for performance reasons.

* To generate text from the status bar object, retrieve the text of each status bar child using the IAccessibleText interface. (The default implementation in api.py only uses the object names and values, which are both empty.)

### Testing strategy:

1) start NVDA
2) start LibreOffice Writer
3) Press NVDA+End key to trigger announcement of status line 4) check that the visible content of the status bar is announced
   (e.g. something like "Page 1 of 1 5 words, 20 characters  Default
   Page Style English (United States) {en-US}       80%")
5) repeat the above steps, but use LibreOffice Calc, Draw and Impress
   instead of Writer (in step 2)

### Known issues with pull request:

none

### Change log entries:

Changes
`Announcement of the status bar (e.g. triggered
by NVDA+End) works for LibreOffice. (#11698)`

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

File: source/appModules/soffice.py [hunk 1]
```diff
@@ -340,3 +340,30 @@ def event_NVDAObject_init(self, obj):
 			# This is a word processor document.
 			obj.description = None
 			obj.treeInterceptorClass = SymphonyDocument
+
+	def searchStatusBar(self, obj, max_depth=5):
```
Comment (seanbudd) [inline source/appModules/soffice.py:344]: could all these added functions please get type hints added to the headers?

For example:
```suggestion
	def searchStatusBar(self, obj: NVDAObject, max_depth: int = 5) -> Optional[NVDAObject]:
```

File: source/appModules/soffice.py [hunk 2]
```diff
@@ -340,3 +340,30 @@ def event_NVDAObject_init(self, obj):
...
 			obj.description = None
 			obj.treeInterceptorClass = SymphonyDocument
+
+	def searchStatusBar(self, obj, max_depth=5):
+		"""Searches for and returns the status bar object
+		if either the object itself or one of its recursive children
+		(up to the given depth) has the corresponding role."""
+		if obj.role == controlTypes.Role.STATUSBAR:
+			return obj
+		if max_depth < 1 or obj.role not in [controlTypes.Role.ROOTPANE, controlTypes.Role.WINDOW]:
```
Comment (seanbudd) [inline source/appModules/soffice.py:350]: A set is probably more appropriate here, not that it is a big deal

```suggestion
		if max_depth < 1 or obj.role not in {controlTypes.Role.ROOTPANE, controlTypes.Role.WINDOW}:
```

**Review by seanbudd — COMMENTED** [review]
Thanks @michaelweghorn, the approach looks good, just need type hints added to each function

File: source/appModules/soffice.py [hunk 3]
```diff
@@ -17,9 +17,11 @@
 import textInfos
 import colors
 from compoundDocuments import CompoundDocument, TreeCompoundTextInfo
+from NVDAObjects import NVDAObject
 from NVDAObjects.IAccessible import IAccessible, IA2TextTextInfo
 from NVDAObjects.behaviors import EditableText
 from logHandler import log
+from typing import Optional
```
Comment (seanbudd) [inline source/appModules/soffice.py:24]: can this be added to the typing import at the start of the file?

File: source/appModules/soffice.py [hunk 4]
```diff
@@ -340,3 +342,30 @@ def event_NVDAObject_init(self, obj):
...
+			return obj
+		if max_depth < 1 or obj.role not in {controlTypes.Role.ROOTPANE, controlTypes.Role.WINDOW}:
+			return None
+		for child in obj.children:
+			status_bar = self.searchStatusBar(child, max_depth - 1)
+			if status_bar:
+				return status_bar
+		return None
+
+	def _get_statusBar(self):
```
Comment (seanbudd) [inline source/appModules/soffice.py:360]: Added a return type for this

```suggestion
	def _get_statusBar(self) -> Optional[NVDAObject]:
```

Comment (michaelweghorn) [reply source/appModules/soffice.py]: Done

**Review by seanbudd — APPROVED** [review]
thanks @michaelweghorn

**Pull request merged**